tutorials/intermediate/blitting.py

```diff
@@ -0,0 +1,186 @@
+# callback to capture a new background whenever a full re-draw
+# happens.
+
+
+class BlitManager:
+    def __init__(self, canvas, animated_artists=()):
+        """
+        Parameters
+        ----------
+        canvas : CanvasAgg
+            The canvas to work with, this only works for sub-classes of the Agg
+            canvas which have the `~CanvasAgg.copy_from_bbox` and
+            `~CanvasAgg.restore_region` methods.
+
+        animated_artists : Iterable[Artist]
+            List of the artists to manage
+        """
+        self.canvas = canvas
+        self._bg = None
+        self._artists = []
+
+        for a in animated_artists:
+            self.add_artist(a)
+        # grab the background on every draw
+        self.cid = canvas.mpl_connect("draw_event", self.on_draw)
+
+    def on_draw(self, event):
+        """Callback to register with 'draw_event'."""
+        cv = self.canvas
+        if event is not None:
+            if event.canvas != cv:
+                raise RuntimeError
+        self._bg = cv.copy_from_bbox(cv.figure.bbox)
+        self._draw_animated()
+
+    def add_artist(self, art):
+        """
+        Add an artist to be managed.
+
+        Parameters
+        ----------
+        art : Artist
+
+            The artist to be added.  Will be set to 'animated' (just
+            to be safe).  *art* must be in the figure associated with
+            the canvas this class is managing.
+
+        """
+        if art.figure != self.canvas.figure:
+            raise RuntimeError
+        art.set_animated(True)
+        self._artists.append(art)
+
+    def _draw_animated(self):
+        """Draw all of the animated artists."""
+        fig = self.canvas.figure
+        for a in self._artists:
+            fig.draw_artist(a)
+
+    def update(self):
+        """Update the screen with animated artists."""
+        cv = self.canvas
+        fig = cv.figure
+        # paranoia in case we missed the draw event,
+        if self._bg is None:
+            self.on_draw(None)
+        else:
+            # restore the old background
+            cv.restore_region(self._bg)
+            # draw all of the animated artists
+            self._draw_animated()
+            # update the screen
+            cv.blit(fig.bbox)
+        # let the GUI event loop process anything it has to do
+        cv.flush_events()
+
+
+###############################################################################
+# And now use our class.  This is a slightly more complicated example
+# of the first case as we add a text frame counter as well.
+
+# make a new figure
+fig, ax = plt.subplots()
+# add a line
+(ln,) = ax.plot(x, np.sin(x), animated=True)
+# add a frame number
+fr_number = ax.annotate(
+    "0",
+    (0, 1),
+    xycoords="axes fraction",
+    xytext=(10, -10),
+    textcoords="offset points",
+    ha="left",
+    va="top",
+    animated=True,
+)
+bm = BlitManager(fig.canvas, [ln, fr_number])
+plt.pause(0.1)
```

Comment:
Maybe add a comment like "plt.pause can go anywhere before the first call to bm.update"? 99% of people are going to skip the huge class definition and just straight to the example so they can copy/paste the class and use it themselves (assuming they made it this far, they're probably either just curious or need to do something the Animation classes are giving them trouble with?)

Reply:
Probably should replace this with lower-level things to get the figure on the screen.